### PR TITLE
Introduce annotateError function for adding telemetry props to an arbitrary error object

### LINKED ISSUE #6573: Replace LoggingError.is with isILoggingError
We shouldn't ever care about whether an error is an instance of LoggingError (via __isFluidLoggingError__ prop).  Instead we just care if it implements the interface.

One minor complication is that the interface is lacking the important `addTelemetryProperties` method and is in `common-definitions` so won't be easily updated.  So we might need a stop-gap in telemetry-utils or something that confirms it has the full interface shape even before we pick up an updated common-definitions.

(And we should update the interface in common-definitions too)

### LINKED ISSUE #6748: Use annotateError instead of relying on LoggingError.addTelemetryProperties directly
This is a more flexible approach that can handle any arbitrary input, and lets us stop relying on an implementation detail (does this object inherit from `LoggingError`?) to be able to annotate it with logging properties.

### PR DESCRIPTION
**Pull request by markfields**
This simplifies the process of annotating errors with logging props - we no longer need to keep track of whether an error is a `LoggingError` - `annotateError` can handle any arbitrary input.

Still need to write/fix tests.

Also fixes #6573

### CODE REVIEW

File: packages/utils/telemetry-utils/src/logger.ts [hunk 1]
```diff
@@ -529,6 +529,43 @@ export function isTaggedTelemetryPropertyValue(x: any): x is ITaggedTelemetryPro
...
+        return error;
+    }
+
+    if (isRegularObject(error)) {
+        // Even though it's not exposed, fully implement IRwLoggingError for subsequent calls to annotateError
+        const loggingError = error as RwLoggingError;
+
+        const propsForError = {...props};
+        loggingError.getTelemetryProperties = () => propsForError;
+        loggingError.addTelemetryProperties = (p: ITelemetryProperties) => { Object.assign(propsForError, p); };
```
Comment (anthony-murphy) [inline packages/utils/telemetry-utils/src/logger.ts:561]: annotations will override existing properties which i'm not sure is correct

File: packages/utils/telemetry-utils/src/logger.ts [hunk 2]
```diff
@@ -529,6 +529,43 @@ export function isTaggedTelemetryPropertyValue(x: any): x is ITaggedTelemetryPro
...
+        return loggingError;
+    }
+
+    const message = String(error);
+    return new LoggingError(message, props);
```
Comment (anthony-murphy) [inline packages/utils/telemetry-utils/src/logger.ts:566]: can we remove the create case from this method, or have one that doesn't have it? errors should always be objects anyway

File: packages/test/test-end-to-end-tests/src/test/error.spec.ts [hunk 3]
```diff
@@ -99,12 +100,15 @@ describeNoCompat("Errors Types", (getTestObjectProvider) => {
...
+    //* test CreatecontainerError around errorType precedence, and whether the error is LoggingError or not
+    //* test createOdspNetworkError around the props that should be added
+
     it("GeneralError Logging Test", async () => {
         const err = {
             userData: "My name is Mark",
             message: "Some message",
         };
-        const iError = (CreateContainerError(err) as any) as LoggingError;
+        const iError = (CreateContainerError(err) as any) as ILoggingError;
```
Comment (anthony-murphy) [inline packages/test/test-end-to-end-tests/src/test/error.spec.ts:111]: you've bundled a lot of refactoring in this PR, and we split that out into a different one

File: packages/utils/telemetry-utils/src/logger.ts (hunk 1 above)
Comment (markfields) [reply packages/utils/telemetry-utils/src/logger.ts:561]: Actually I'm going to keep it as-is for now, to match `LoggingError.addTelemetryProperties`.  I will submit a follow-up PR to flip them both.

**Review by anthony-murphy — APPROVED** [review]
:shipit:

File: packages/utils/telemetry-utils/src/logger.ts (hunk 2 above)
Comment (markfields) [reply packages/utils/telemetry-utils/src/logger.ts:567]: For context, here's a preview of the next change I'll make, to enhance `annotateError` to yield a proper `IFluidErrorBase`:

https://github.com/markfields/FluidFramework/pull/1

**Comment by markfields** [comment]
Note that after further discussion with Tony and Vlad, I mostly reverted this as part of #6764

**Pull request merged**